datafusion/core/src/physical_optimizer/enforce_sorting.rs

```diff
@@ -131,7 +131,7 @@ fn update_coalesce_ctx_children(
     coalesce_context.data = if children.is_empty() {
         // Plan has no children, it cannot be a `CoalescePartitionsExec`.
         false
-    } else if is_coalesce_partitions(&coalesce_context.plan) {
+    } else if is_coalesce_partitions(&coalesce_context.plan) || coalesce_context.data {
```

Comment:
I think none of the changes in this file is necessary for this PR. I retracted changes in this file. All tests still pass